### ISSUE TITLE
Bug: [unbound-method] Reword message to better explain `this: void` binding

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10056]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=MYewdgzgLgBA3jAhgIxAJ1gXxgXhmAUwHcYBBVDAYXCjRABt6C0AKASgG4AoLldKdhyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const { abort } = new AbortController();

abort();
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The error message should tell me what's wrong, and optionally give a hint as to how _I_ can fix it. If the type comes from a `.d.ts` outside of my control -say, `node_modules/` or `lib.*.d.ts`- then there's no need to suggest changes to the type signature such as annotating with `this: void`.

### Actual Result

```plaintext
Avoid referencing unbound methods which may cause unintentional scoping of `this`.
If your function does not access `this`, you can annotate it with `this: void`, or consider using an arrow function instead. 
```

### Additional Info

Maybe we just remove the _"If your ... annotate it with `this: void`"_ from the error? 

```plaintext
Avoid referencing unbound methods which may cause unintentional scoping of `this`.
Consider using an arrow function instead. 
```

Thanks @mattpocock for accidentally reminding me to file this, I'd been meaning to for a while. 😄 https://x.com/JoshuaKGoldberg/status/1838944653401375048

💖

**Comment by kirkwaiblinger** [comment]
My feeling is slightly negative on the general principle of inferring whether something can be changed based on where it's declared, and creating a conditional error message.

Things in node_modules may very well be under the user's control if they are importing another package they have control over. `lib.*.d.ts`.... less flexible of course.

_At the call site_, your only real resolution to this rule is to use an arrow function or do `x.bind(x)`. But it's useful and relevant info to know that the declaration is at fault if you think the report is wrong (then again who cares - just use an arrow function anyway and move on).

Personally I think I slightly prefer removing the part about `this: void` unconditionally from the error message and keeping it on the docs page, over attempting to infer whether a symbol's declaration is something the user has power to change.

***

Another possibility would be to reword the error message to say something like
```text
A method that is not declared with `this: void` may cause unintentional scoping of `this` when separated from its object.
Consider using an arrow function or explicitly `.bind()`ing the method to avoid calling the method with an unintended `this` value. 
```

In other words _because of the way this function is declared_, it's a problem to use it this way. You can use it safely it these other ways. (But you can also infer that if you can change how the function is declared you can also prevent the report in teh first place)

**Comment by valler** [comment]
Can someone explain why this rule errors on the call site?

This already bit me more than once with false positives when working with external packages. My thoughts on this are as follows.

1) The suggestion to let the caller bind or wrap any unbound function or method for usage as a callback kinda advocates code smell (one implementation, up to infinity call sites), while passing around functions is perfectly idiomatic JS/TS (debatable).
2) Giving the call site tipps about improvements for the callee doesn't help at all, because as already mentioned, one might not be in control.
3) The caller should only be concerned about matching signatures and passing correct arguments. Most certainly the caller should not care about where and how a callable was defined, other than constructables (`new`), because of (1) and because there's no way to tell or know, other than reading the source.
4) The only scenario I can think of, where the current behaviour could apply, is when the implementation *actually* uses `this` without binding (assuming leaking implementation details are of no concern).

The linter should error on the callee. The function/method implementor should be aware of (1) and the linter should suggest implementation patterns which express clear intent and keep the call site safe from errors and workarounds to allow for (3). Even if this proves to be impossible to achieve in terms of language syntax, other than comments, it would be better to have these linter comments and errors on the callee's site (but `this: void` always works, does it?).

**Comment by kirkwaiblinger** [comment]
@valler 

without some code examples, I'm having a hard time following https://github.com/typescript-eslint/typescript-eslint/issues/10056#issuecomment-2467053331. Also, this issue is just about the wording of the report message, not about where the rule reports.

To help us out, could you do one of the following, whichever is most appropriate
- file a new issue with specific code examples and suggested changes to the rule's behavior
- reach out to us on discord to ask more open-ended questions (see [contributing guide](https://typescript-eslint.io/contributing/issues/#questions-and-support-requests))
- comment on this issue with some code examples to support https://github.com/typescript-eslint/typescript-eslint/issues/10056#issuecomment-2467053331, and explain its relevance to this issue

**Comment by valler** [comment]
Thanks @kirkwaiblinger

I opened #10318, since only (2) in my original comment was directly related to the wording of the error message.

**Comment by valler** [comment]
Back at this issue, because

- according to https://github.com/typescript-eslint/typescript-eslint/issues/10318 this rule is working as intended and, overall, I agree
- in https://github.com/microsoft/TypeScript/issues/60477 commenters call it a false positive, despite the lack of `this: void` annotations or alternative solutions
- as the title of this issue suggests, these errors primarily stem from external code with different linter settings, or at least these are the ones the unbinding site has no control over fixing where they actually should be fixed, which makes it hard to get rid of the error.

In conclusion I assert, contrary to some comments in https://github.com/typescript-eslint/typescript-eslint/issues/10318, unbinding to pass callbacks is indeed idiomatic JS/TS, regardless of arrow function usage.

However, I *disagree* with the idea of a message that encourages "fixing" by means of `.bind(...)` etc. As of now, the message is perfectly fine with suggesting fixes where the function is *defined*. I can confirm though, that even without such an explicit suggestion, I was inclined to wrapping "false positives" in arrow functions, just to get rid of the message. If anything, one should steer the programmer away from the wrong kind of "fix". If the error is not fixable on your end, first make sure its actually an error (know what you're importing). Commenting out this rule may very well be the better solution, over `bind` and wrappers which do nothing but pass through arguments.

I also *disagree* with silencing the message for external code. A (potential) error puts you at risk, regardless of where it occurs. Third party code is still a part of your app. It's important to be aware of what could happen at runtime.

If not in the message itself, maybe it's possible to address this in the docs? If not, I'd rather change nothing about it.

**Comment by valler** [comment]
Shouldn't the linter respect `skipLibCheck` and related settings? This should make the problem go away, right?

**Comment by bradzacher** [comment]
The core of the issues is that there's currently no generic way for the rule to tell whether or not a function must be bound.

Even if we had logic to skip across function boundaries and deeply analyse every single function - we'd get stuck at 3rd party code because such code is shipped as `.js` files with `.d.ts` signatures and the only thing TS shows us is the `.d.ts` signatures and provides no mechanism to get to the implementation.

With that in mind there's absolutely no way to generically tell if a function uses `this` from its signature alone because TS does not enforce a `this` parameter.

Additionally in TS there isn't any differentiation between a class and an object at a type-system level. They're both represented in the same way - as an object type. So we can't  refine this assumption to only apply to "class-like" things because there's no such thing. Also it's completely valid to use `this` in a non-class object too - meaning even if we could such an assumption wouldn't technically be valid.

All of this means that for the rule to exist it must make assumptions - the assumption that it makes is that any method signature is a bound function and a function type is not.

Unfortunately this assumption does not always hold true and can have false positives - especially with 3rd party code. There isn't really a solution to make it more accurate.
